Resolving add-on references for backwards compatibility and better UX

bjeanes opened the pull request:
~~I need to break this up into smaller commits as I am doing too much at once in the current version here.~~ (update: done)

~~Nonetheless, I'm sticking it up here because it's at a reasonably good point and wouldn't mind some feedback on the direction (or code, but that's less final).~~

Basically, this makes more of the CLI commands (`docs`, `open`, soon `attach`, `detach` and `destroy`) be able to resolve an add-on reference (a la #34) but also improves the resolution logic itself (`HEROKU_POSTGRESQL_CYAN` on `some-app`/ `some-app::HEROKU_POSTGRESQL_CYAN` [or any substring of the attachment name, as is common with the `pg:*` commands], `[@]resource-name`, `service[:plan]`, or UUID all work for add-on identification).  ~~Lastly, it just pulls a lot of helper methods into helper modules to make the `Command` module very focussed on the actual CLI portions. The goal of this extraction is to get closer towards being able to unit test some of this logic, but I still don't see an easy way to test the plugin as a whole. Ideas very welcome here.~~ (edit: moved to other PRs and merged)

attn: @heroku/add-ons

mattreduce commented:
@bjeanes this is looking great. Thanks for also extracting all of those API calls we had everywhere... much better

bjeanes commented:
@mattreduce I'll probably spend some time today extracting out the API-only changes because they'll be easy to review and leave the rest of the changes easier to comprehend.

bjeanes commented:
@mattreduce OK I've split this up into some more logical commits now. This should be a bit easier to review. I'll probably also move some of these into separate PRs to merge earlier, where I can.

mattreduce commented:
@bjeanes nice, thanks! I'll dig into this again and compare it with what I've got for config vars, adjust where needed

bjeanes commented:
@mattreduce this is ready for a review now. I split out the other work I did into other PRs (some of which I merged because they were simple enough and I tested them pretty thoroughly). This diff should be pretty focused now.

bjeanes commented on lib/ext/heroku/helpers/addons/resolve.rb:
```diff
@@ -0,0 +1,127 @@
+      def resolve_attachment(identifier)
+        case identifier
+        when UUID
+          [get_attachment(identifier)]
```
might need a `.compact` here, now that I think about it

mattreduce replied:
Ah yes, otherwise it could technically be an array of one element, `nil`

bjeanes commented on lib/ext/heroku/helpers/addons/resolve.rb:
```diff
@@ -0,0 +1,127 @@
+        end
+      end
+
+      # Resolve unique add-on or return error using:
```
Note to self: out-of-date comment. This method returns an Array.

mattreduce replied:
:ok_hand:

mattreduce commented:
This looks good to me. I didn't see any problems, but we don't have tests wrapped around all of this so it's hard to say. If it's all functional I'm happy to merge it :+1:

bjeanes commented:
@mattreduce addressed the comments and generally cleaned up the code comments. Merge if you're comfortable, or feel free to play with it a bit first or cover with direwolf. I want to get this in front of people soon, obviously, but not as much as I want to make sure it *works*.

Last night I ran through each of the commands manually with as many edge cases as I could think of on the spot, but this is far from thorough.

bjeanes commented:
@mattreduce and other reviewers...

Add this as a command to the toolbelt, for quick and easy manual testing of the resolution logic itself:

``` diff
+
+    def resolve
+      styled_hash(resolve_addon!(shift_argument))
+    end
```

Example:

``` sh-session
$ heroku addons:resolve guarded-eyrie-2616::GREEN # by substring attachment with app prefix
addon_service: {"id"=>"6c67493d-8fc2-4cd4-9161-4f1ec11cbe69", "name"=>"heroku-postgresql"}
app:           {"id"=>"97be3ac6-2bdc-416f-85ff-009dc6f21f5d", "name"=>"guarded-eyrie-2616"}
config_vars:   HEROKU_POSTGRESQL_GREEN_URL
created_at:    2015-04-07T22:49:29Z
id:            5d089653-ea4f-4f26-a277-472adde72ad2
name:          askfjshfjashdflaksdjhfaksjdfhasdjkfx1
plan:          {"created_at"=>"2014-08-07T20:34:30Z", "default"=>false, "description"=>"Heroku Postgres Standard 0", "human_name"=>"Standard 0", "id"=>"112f476d-536b-40a9-9d47-ce346920755d", "name"=>"heroku-postgresql:standard-0", "price"=>{"cents"=>5000, "unit"=>"month"}, "state"=>"ga", "updated_at"=>"2014-09-04T14:03:26Z"}
provider_id:   resource8959341@heroku.com
updated_at:    2015-04-07T22:49:34Z
web_url:       https://postgres.heroku.com/databases/guarded-eyrie-2616/heroku-postgresql-green

$ heroku addons:resolve heroku-postgresql # by ambiguous service
 !    Multiple add-ons match "heroku-postgresql".
 !    Use the name of add-on resource:
 !
 !    - askfjshfjashdflaksdjhfaksjdfhasdjkfx1 (heroku-postgresql:standard-0)
 !    - budding-purposefully-5254 (heroku-postgresql:hobby-dev)

$ heroku addons:resolve heroku-postgresql:hobby-dev # by unambiguous service:plan
addon_service: {"id"=>"6c67493d-8fc2-4cd4-9161-4f1ec11cbe69", "name"=>"heroku-postgresql"}
app:           {"id"=>"97be3ac6-2bdc-416f-85ff-009dc6f21f5d", "name"=>"guarded-eyrie-2616"}
config_vars:   DATABASE_URL
created_at:    2015-04-08T03:52:39Z
id:            3c752317-2bb4-4f3e-bdb7-b4dcca064caa
name:          budding-purposefully-5254
plan:          {"created_at"=>"2013-10-24T17:21:05Z", "default"=>true, "description"=>"Heroku Postgres Hobby Dev", "human_name"=>"Hobby Dev", "id"=>"062a1cc7-f79f-404c-9f91-135f70175577", "name"=>"heroku-postgresql:hobby-dev", "price"=>{"cents"=>0, "unit"=>"month"}, "state"=>"ga", "updated_at"=>"2014-06-18T17:06:29Z"}
provider_id:   9090257
updated_at:    2015-04-08T03:52:40Z
web_url:       https://postgres.heroku.com/databases/guarded-eyrie-2616/database
```
